File: src/prod/src/ServiceModel/CodePackageEntryPoint.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;
using namespace Common;
using namespace ServiceModel;

CodePackageEntryPoint::CodePackageEntryPoint()
    : entryPointLocation_()
    , processId_()
    , runAsUserName_()
    , nextActivationTime_(DateTime::Zero)
    , entryPointStatus_(EntryPointStatus::Invalid)
    , statistics_()
    , instanceId_(0)
{
}

CodePackageEntryPoint::CodePackageEntryPoint(
    wstring const & entryPointLocation, 
    int64 processId, 
    wstring const & runAsUserName, 
    DateTime const & nextActivationTime,        
    EntryPointStatus::Enum entryPointStatus,
    CodePackageEntryPointStatistics const & statistics,
    int64 instanceId)
    : entryPointLocation_(entryPointLocation)
    , processId_(processId)
    , runAsUserName_(runAsUserName)
    , nextActivationTime_(nextActivationTime)
    , entryPointStatus_(entryPointStatus)
    , statistics_(statistics)
    , instanceId_(instanceId)
{
}

void CodePackageEntryPoint::ToPublicApi(
    __in Common::ScopedHeap & heap, 
    __out FABRIC_CODE_PACKAGE_ENTRY_POINT & publicCodePackageEntryPoint) const
{
    publicCodePackageEntryPoint.EntryPointLocation = heap.AddString(entryPointLocation_);
    publicCodePackageEntryPoint.ProcessId = processId_;
    publicCodePackageEntryPoint.RunAsUserName = heap.AddString(runAsUserName_);
    publicCodePackageEntryPoint.NextActivationUtc = nextActivationTime_.AsFileTime;
    publicCodePackageEntryPoint.EntryPointStatus = EntryPointStatus::ToPublicApi(entryPointStatus_);

    auto publicCodePackageStatistics = heap.AddItem<FABRIC_CODE_PACKAGE_ENTRY_POINT_STATISTICS>();
    statistics_.ToPublicApi(heap, *publicCodePackageStatistics);    
    publicCodePackageEntryPoint.Statistics = publicCodePackageStatistics.GetRawPointer();

    auto publicCodePackageEntryEx1 = heap.AddItem<FABRIC_CODE_PACKAGE_ENTRY_POINT_EX1>();
    publicCodePackageEntryEx1->CodePackageInstanceId = instanceId_;

    publicCodePackageEntryPoint.Reserved = publicCodePackageEntryEx1.GetRawPointer();
}

void CodePackageEntryPoint::WriteTo(Common::TextWriter& w, Common::FormatOptions const &) const
{
    w.Write(ToString());
}

wstring CodePackageEntryPoint::ToString() const
{
    return wformatString(
        "EntryPointLocation = {0}, ProcessId = {1}, RunAsUserName = {2}, NextActivationTime = {3}, Status = {4}, Statistics = [{5}], InstanceId = [{6}]",
        entryPointLocation_,
        processId_,
        runAsUserName_,
        nextActivationTime_,
        entryPointStatus_,
        statistics_.ToString(),
        instanceId_);
}

ErrorCode CodePackageEntryPoint::FromPublicApi(__in FABRIC_CODE_PACKAGE_ENTRY_POINT const & publicDeployedCodePackageQueryResult)
{
    entryPointLocation_ = wstring(publicDeployedCodePackageQueryResult.EntryPointLocation);
    processId_ = publicDeployedCodePackageQueryResult.ProcessId;
    runAsUserName_ = wstring(publicDeployedCodePackageQueryResult.RunAsUserName);
    nextActivationTime_ = DateTime(publicDeployedCodePackageQueryResult.NextActivationUtc);

    auto error = EntryPointStatus::FromPublicApi(publicDeployedCodePackageQueryResult.EntryPointStatus, entryPointStatus_);
    if(!error.IsSuccess()) { return error; }

    if (publicDeployedCodePackageQueryResult.Reserved != NULL)
    {
        FABRIC_CODE_PACKAGE_ENTRY_POINT_EX1 * publicCodePackageEntryResultEx1 = (FABRIC_CODE_PACKAGE_ENTRY_POINT_EX1*) publicDeployedCodePackageQueryResult.Reserved;
        instanceId_ = publicCodePackageEntryResultEx1->CodePackageInstanceId;
    }

    return statistics_.FromPublicApi(*publicDeployedCodePackageQueryResult.Statistics);
}

File: src/prod/src/managed/Api/src/System/Fabric/query/CodePackageEntryPoint.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Query
{
    using System.Fabric.Interop;
    using System.Fabric.Common.Serialization;

    /// <summary>
    /// <para>Represents a code package entry point.</para>
    /// </summary>
    public sealed class CodePackageEntryPoint
    {
        internal CodePackageEntryPoint(
            string entryPointLocation,
            long processId,
            string runAsUserName,
            EntryPointStatus entrypointStatus,
            DateTime nextActivationUtc,
            CodePackageEntryPointStatistics statistics,
            Int64 codePackageInstanceId)
        {
            this.EntryPointLocation = entryPointLocation;
            this.ProcessId = processId;
            this.RunAsUserName = runAsUserName;
            this.EntryPointStatus = entrypointStatus;
            this.NextActivationUtc = nextActivationUtc;
            this.Statistics = statistics;
            this.CodePackageInstanceId = codePackageInstanceId;
        }

        private CodePackageEntryPoint() { }

        /// <summary>
        /// <para>Gets the entry point location.</para>
        /// </summary>
        /// <value>
        /// <para>The entry point location.</para>
        /// </value>
        public string EntryPointLocation { get; private set; }

        /// <summary>
        /// <para>Gets the process ID.</para>
        /// </summary>
        /// <value>
        /// <para>The process ID.</para>
        /// </value>
        public long ProcessId { get; private set; }

        /// <summary>
        /// <para>Gets the run as user name.</para>
        /// </summary>
        /// <value>
        /// <para>The run as user name.</para>
        /// </value>
        public string RunAsUserName { get; private set; }

        /// <summary>
        /// <para>Gets the statistics.</para>
        /// </summary>
        /// <value>
        /// <para>The statistics.</para>
        /// </value>
        [JsonCustomization(PropertyName = JsonPropertyNames.CodePackageEntryPointStatistics)]
        public CodePackageEntryPointStatistics Statistics { get; private set; }

        /// <summary>
        /// <para>Gets the entry point status.</para>
        /// </summary>
        /// <value>
        /// <para>The entry point status.</para>
        /// </value>
        [JsonCustomization(PropertyName = JsonPropertyNames.Status)]
        public EntryPointStatus EntryPointStatus { get; private set; }

        /// <summary>
        /// <para>Gets the next activation UTC.</para>
        /// </summary>
        /// <value>
        /// <para>The next activation UTC.</para>
        /// </value>
        [JsonCustomization(PropertyName = JsonPropertyNames.NextActivationTime)]
        public DateTime NextActivationUtc { get; private set; }

        /// <summary>
        /// <para>
        /// Gets the code package instance id.
        /// </para>
        /// </summary>
        /// <value>
        /// <para>The code package instance id.</para>
        /// </value>
        [JsonCustomization(PropertyName = JsonPropertyNames.InstanceId)]
        public Int64 CodePackageInstanceId { get; private set; }

        internal static unsafe CodePackageEntryPoint FromNative(NativeTypes.FABRIC_CODE_PACKAGE_ENTRY_POINT nativeEntryPoint)
        {
            NativeTypes.FABRIC_CODE_PACKAGE_ENTRY_POINT_STATISTICS* nativeEntryPointStatistics = (NativeTypes.FABRIC_CODE_PACKAGE_ENTRY_POINT_STATISTICS*)nativeEntryPoint.Statistics;

            Int64 instanceId = 0;
            if (nativeEntryPoint.Reserved != IntPtr.Zero)
            {
                NativeTypes.FABRIC_CODE_PACKAGE_ENTRY_POINT_EX1* nativeResultItemEx1 = (NativeTypes.FABRIC_CODE_PACKAGE_ENTRY_POINT_EX1*)nativeEntryPoint.Reserved;
                instanceId = nativeResultItemEx1->CodePackageInstanceId;
            }

            return new CodePackageEntryPoint(
                NativeTypes.FromNativeString(nativeEntryPoint.EntryPointLocation),
                nativeEntryPoint.ProcessId,
                NativeTypes.FromNativeString(nativeEntryPoint.RunAsUserName),
                (EntryPointStatus)nativeEntryPoint.EntryPointStatus,
                NativeTypes.FromNativeFILETIME(nativeEntryPoint.NextActivationUtc),
                CodePackageEntryPointStatistics.FromNative(*nativeEntryPointStatistics),
                instanceId);
        }
    }
}